src/dev/adam_practice_tagfinder.py

```diff
@@ -0,0 +1,66 @@
+import bs4
```

Comment:
Hmmmm you have very complicated code with lots of conversation between dicts and lots of sorting sorting an swapping.

And I think because you only have a nest list you may not be going down far enough into the very deep html we will find on real websites.

If we think about it all we need is a count of how far away from the parent node a given node is (giving us some sort of measure of the 'generation' of every node) and also a count of how many <p> tags are contained in all ancestors.

So we want to end up with two dictionaries like this:

```
node_depth = {'html': 0,
 'head': 1,
 'title': 2,
 'style': 2,
 'random_section': 1,
 'random_body_name_1234': 1,
 'div_first_div': 2,
 'div_second_div': 2,
 'div_third_div': 2,
 'div_fourth_div': 2,
 'div_fifth_div': 2, }

n_p_tags = {'html': 7,
 'head': 0,
 'title': 0,
 'style': 2,
 'random_section': 2,
 'random_body_name_1234': 5,
 'div_first_div': 1,
 'div_second_div': 1,
 'div_third_div': 1,
 'div_fourth_div': 1,
 'div_fifth_div': 1, }
```

If you can construct those dictionaries I think the following code will solve our problems:
(you will need to `pip install pandas` in the terminal)
```
import pandas 
node_depth_df = pd.DataFrame.from_dict(node_depth, orient='index', columns=['node_depth'])#.sort_values(by='depth', ascending=False)

n_p_tags_df = pd.DataFrame.from_dict(n_p_tags, orient='index', columns=['p_tag_count'])#.sort_values(by='count', ascending=False)

all_df = pd.merge(node_depth_df, n_p_tags_df, right_index=True, left_index=True).sort_values(by='node_depth')

depth_times_count_ser = all_df.apply(lambda row: row.node_depth * row.p_tag_count, axis=1)

node_we_need = depth_times_count_ser.idxmax()
```